pint/models/timing_model.py

```diff
@@ -774,14 +792,42 @@ def designmatrix(self, toas, acc_delay=None, scale_by_F0=True, \
             M[:, mask] /= F0.value
         return M, params, units, scale_by_F0
 
-    def read_parfile(self, filename):
-        """Read values from the specified parfile into the model parameters."""
+    def read_parfile(self, parfile):
+        """Read parameter values from a .parfile.
+
+           Parameter
+           ---------
+           parfile: str or .par file line lists
+               .par file name.
+        """
+        if isinstance(parfile, str):
+            pfile = open(parfile, 'r')
+            plines = pfile.readlines()
+        elif isinstance(parfile, (list, tuple)):
+            plines = parfile
+        else:
+            raise ValueError("Acceptable .par file format: `filename` and "
+                             "`.par file line list.`")
+        self.read_parfile_string(plines)
+        if isinstance(parfile, str):
+            pfile.close()
+
+    def read_parfile_string(self, par_strings):
+        """Read values from a parfile in string list format
+           into the model parameters.
+
+           Parameter
+           ---------
+           par_strings: list
+               List of parfile line strings.
+        """
         checked_param = []
         repeat_param = {}
         param_map = self.get_params_mapping()
         comps = self.components
-        pfile = open(filename, 'r')
-        for l in [pl.strip() for pl in pfile.readlines()]:
+        for l in [pl.strip() for pl in par_strings]:
```

Comment:
I recommend (writing and) using a generator that yields only non-comment non-blank stripped lines. I bet this isn't the only place we'd use such a thing. Particularly if the comment marker was a parameter.